src/core/lib/channel/promise_based_filter.h

```diff
@@ -325,39 +374,137 @@ class CallData<ChannelFilter, true> : public BaseCallData {
+              abort();  // unimplemented
+            case RecvTrailingState::kResponded:
+            case RecvTrailingState::kCancelled:
+              abort();  // unreachable
+          }
+          if (destroy_md) {
+            md->~grpc_metadata_batch();
+          }
         }
       } break;
       case SendInitialState::kInitial:
       case SendInitialState::kCancelled:
-        // If we get a response without sending anything, we just propagate that
-        // up. (note: that situation isn't possible once we finish the promise
-        // transition).
+        // If we get a response without sending anything, we just propagate
+        // that up. (note: that situation isn't possible once we finish the
+        // promise transition).
         if (recv_trailing_state_ == RecvTrailingState::kComplete) {
           recv_trailing_state_ = RecvTrailingState::kResponded;
           call_closure =
               absl::exchange(original_recv_trailing_metadata_ready_, nullptr);
         }
         break;
     }
+    GRPC_CALL_STACK_REF(call_stack(), "finish_poll");
     is_polling_ = false;
+    bool in_combiner = true;
+    if (forward_batch != nullptr) {
+      in_combiner = false;
+      forward_send_initial_metadata_ = false;
+      grpc_call_next_op(elem(), forward_batch);
+    }
+    if (cancel_send_initial_metadata_error != GRPC_ERROR_NONE) {
+      GPR_ASSERT(in_combiner);
+      forward_send_initial_metadata_ = false;
+      in_combiner = false;
+      grpc_transport_stream_op_batch_finish_with_failure(
+          absl::exchange(send_initial_metadata_batch_, nullptr),
+          cancel_send_initial_metadata_error, call_combiner());
+    }
     if (absl::exchange(forward_send_initial_metadata_, false)) {
+      GPR_ASSERT(in_combiner);
+      in_combiner = false;
       grpc_call_next_op(elem(),
                         absl::exchange(send_initial_metadata_batch_, nullptr));
     }
     if (call_closure != nullptr) {
+      GPR_ASSERT(in_combiner);
+      in_combiner = false;
       Closure::Run(DEBUG_LOCATION, call_closure, GRPC_ERROR_NONE);
     }
+    if (absl::exchange(repoll_, false)) {
+      if (in_combiner) {
+        WakeInsideCombiner();
+      } else {
+        struct NextPoll : public grpc_closure {
+          grpc_call_stack* call_stack;
+          CallData* call_data;
+        };
+        auto run = [](void* p, grpc_error_handle) {
+          auto* next_poll = static_cast<NextPoll*>(p);
+          next_poll->call_data->WakeInsideCombiner();
+          GRPC_CALL_STACK_UNREF(next_poll->call_stack, "re-poll");
+          delete next_poll;
+        };
+        auto* p = new NextPoll;
+        GRPC_CALL_STACK_REF(call_stack(), "re-poll");
+        GRPC_CLOSURE_INIT(p, run, p, nullptr);
+        GRPC_CALL_COMBINER_START(call_combiner(), p, GRPC_ERROR_NONE,
+                                 "re-poll");
+      }
+    } else if (in_combiner) {
+      GRPC_CALL_COMBINER_STOP(call_combiner(), "poll paused");
```

Comment:
I think I must still be missing something in terms of how the promises interact with the existing filter stack machinery.

Am I correct in understanding that this code will be used only for individual filters that are implemented as promises, not for the case where the entire filter stack has been converted to use promises?  It looks like the activity here is just for this one filter, not for the call as a whole.  If so, doesn't that mean that the promise will only ever be woken up when there's an async callback for some work done *while this filter is processing an op*?

If so, then I don't understand why we ever need to yield or re-enter the call combiner in this code.  The existing filter code is designed to hold the call combiner while a given batch is transiting either down or up through the call stack, even if async work is done during that transit.  It's not clear to me why that model should change just because the async work is now being done via a promise instead of via callbacks.  I think the only time we should need to yield the call combiner would be when the call gets cancelled, and even then it will happen indirectly: I think we should just register a call combiner cancellation callback that cancels the promise and returns a failure for whatever batch we were processing, which will result in yielding the call combiner when that failure reaches the surface.

I will certainly defer to your expertise in performance matters, and even if there is a performance hit here, I do realize that this code is temporary, so it probably doesn't matter in the long run.  But I'm struggling to understand how this approach of always yielding and re-entering the call combiner for async work could *not* cause performance problems, when this is exactly the behavior that I had to work so hard to avoid when I first implemented the call combiner.

What am I missing here?  (Happy to chat offline if that will be easier.)

Reply:
Consider the plugin credentials asynchronous path: we get some op, we call the application, the application says hey I'm not calling back for a while.

In this case we need to drop the combiner (this line) and return from the start op handler.

When we get the application callback, we reacquire the combiner and restart this logic (hopefully pushing the op down the stack)

Reply:
(we don't drop the combiner if, say, we're just pushing an op down the stack and waiting for the callback - that path works as it always has - as indeed does this path)